From ab5400040fcf3a1061229afb3ab81f26fef0858d Mon Sep 17 00:00:00 2001
From: Michael Orlitzky <michael@orlitzky.com>
Date: Fri, 12 Nov 2021 21:33:39 -0500
Subject: Trac #32867: don't export MAXIMA_PREFIX.

This should work automatically, and the previous behavior is
incorrect if we intend to use the system copy of maxima.
---
 src/bin/sage-env | 3 ---
 1 file changed, 3 deletions(-)

diff --git a/src/bin/sage-env b/src/bin/sage-env
index a0426df..f0866b4 100644
--- a/src/bin/sage-env
+++ b/src/bin/sage-env
@@ -482,9 +482,6 @@ if [ -d "$DOT_SAGE" ] ; then
     fi
 fi
 
-if [ -n "$SAGE_LOCAL" ]; then
-    export MAXIMA_PREFIX="$SAGE_LOCAL"
-fi
 export MAXIMA_USERDIR="$DOT_SAGE/maxima"
 
 if [ -n "$SAGE_LOCAL" ]; then
-- 
cgit v1.0-1-gd88e


From 0e29761fbd23736d8d9acd235e4fa2de839df6a4 Mon Sep 17 00:00:00 2001
From: Michael Orlitzky <michael@orlitzky.com>
Date: Fri, 12 Nov 2021 21:13:21 -0500
Subject: Trac #32867: remove the MAXIMA sage_conf variable.

Until now, the MAXIMA variable held the name of the main maxima
executable. But there really is no reason to use anything other
than "maxima -l ecl", which works in all cases. This commit
replaces the variable with the hard-coded value. This simplifies
the use of maxima from the system because we no longer need to
set sage_conf.MAXIMA to any particular value during ./configure.
---
 pkgs/sage-conf/sage_conf.py.in         | 2 --
 src/bin/sage                           | 6 +-----
 src/sage/env.py                        | 1 -
 src/sage/interfaces/maxima.py          | 3 +--
 src/sage/interfaces/maxima_abstract.py | 8 ++++----
 5 files changed, 6 insertions(+), 14 deletions(-)

diff --git a/pkgs/sage-conf/sage_conf.py.in b/pkgs/sage-conf/sage_conf.py.in
index fbce694..b40cda6 100644
--- a/pkgs/sage-conf/sage_conf.py.in
+++ b/pkgs/sage-conf/sage_conf.py.in
@@ -9,8 +9,6 @@ VERSION = "@PACKAGE_VERSION@"
 SAGE_LOCAL = "@prefix@"
 SAGE_ROOT = "@SAGE_ROOT@"
 
-MAXIMA = "@prefix@/bin/maxima"
-
 # Delete this line if your ECL can load maxima without further prodding.
 MAXIMA_FAS = "@SAGE_MAXIMA_FAS@".replace('${prefix}', SAGE_LOCAL)
 
diff --git a/src/bin/sage b/src/bin/sage
index fcd2630..54b69a1 100755
--- a/src/bin/sage
+++ b/src/bin/sage
@@ -684,11 +684,7 @@ fi
 
 if [ "$1" = '-maxima' -o "$1" = '--maxima' ]; then
     shift
-    maxima_cmd=$(sage-config MAXIMA 2>/dev/null)
-    if [ -z "${maxima_cmd}" ]; then
-        maxima_cmd="maxima -l ecl"
-    fi
-    exec $maxima_cmd "$@"
+    exec maxima -l ecl "$@"
 fi
 
 if [ "$1" = '-mwrank' -o "$1" = '--mwrank' ]; then
diff --git a/src/sage/env.py b/src/sage/env.py
index 40ace13..95edc16 100644
--- a/src/sage/env.py
+++ b/src/sage/env.py
@@ -207,7 +207,6 @@ MATHJAX_DIR = var("MATHJAX_DIR", join(SAGE_SHARE, "mathjax"))
 MTXLIB = var("MTXLIB", join(SAGE_SHARE, "meataxe"))
 THREEJS_DIR = var("THREEJS_DIR", join(SAGE_SHARE, "threejs-sage"))
 PPLPY_DOCS = var("PPLPY_DOCS", join(SAGE_SHARE, "doc", "pplpy"))
-MAXIMA = var("MAXIMA", "maxima")
 MAXIMA_FAS = var("MAXIMA_FAS")
 KENZO_FAS = var("KENZO_FAS")
 SAGE_NAUTY_BINS_PREFIX = var("SAGE_NAUTY_BINS_PREFIX", "")
diff --git a/src/sage/interfaces/maxima.py b/src/sage/interfaces/maxima.py
index fc7f48b..7313c7d 100644
--- a/src/sage/interfaces/maxima.py
+++ b/src/sage/interfaces/maxima.py
@@ -484,7 +484,6 @@ import shlex
 
 from random import randrange
 
-from sage.env import MAXIMA
 from sage.misc.misc import ECL_TMP
 
 from .expect import (Expect, ExpectElement, gc_disabled)
@@ -558,7 +557,7 @@ class Maxima(MaximaAbstract, Expect):
         Expect.__init__(self,
                         name = 'maxima',
                         prompt = r'\(\%i[0-9]+\) ',
-                        command = '{0} -p {1}'.format(MAXIMA, shlex.quote(STARTUP)),
+                        command = 'maxima -l ecl -p {0}'.format(shlex.quote(STARTUP)),
                         env = {'TMPDIR': str(ECL_TMP)},
                         script_subdirectory = script_subdirectory,
                         restart_on_ctrlc = False,
diff --git a/src/sage/interfaces/maxima_abstract.py b/src/sage/interfaces/maxima_abstract.py
index 917059d..c966a3c 100644
--- a/src/sage/interfaces/maxima_abstract.py
+++ b/src/sage/interfaces/maxima_abstract.py
@@ -54,7 +54,7 @@ import re
 import sys
 import subprocess
 
-from sage.env import DOT_SAGE, MAXIMA
+from sage.env import DOT_SAGE
 COMMANDS_CACHE = '%s/maxima_commandlist_cache.sobj' % DOT_SAGE
 
 from sage.cpython.string import bytes_to_str
@@ -164,7 +164,7 @@ class MaximaAbstract(ExtraTabCompletion, Interface):
             -- Function: gcd (<p_1>, <p_2>, <x_1>, ...)
             ...
         """
-        cmd = '{} --very-quiet --batch-string="{}({});" '.format(MAXIMA, command, s)
+        cmd = 'maxima -l ecl --very-quiet --batch-string="{}({});" '.format(command, s)
         env = os.environ.copy()
         env['TMPDIR'] = str(ECL_TMP)
 
@@ -2214,7 +2214,7 @@ def maxima_version():
         sage: maxima_version()  # random
         '5.41.0'
     """
-    with os.popen('{} --version'.format(MAXIMA)) as p:
+    with os.popen('maxima -l ecl --version') as p:
         return p.read().split()[-1]
 
 
@@ -2232,4 +2232,4 @@ def maxima_console():
     from sage.repl.rich_output.display_manager import get_display_manager
     if not get_display_manager().is_in_terminal():
         raise RuntimeError('Can use the console only in the terminal. Try %%maxima magics instead.')
-    os.system('{}'.format(MAXIMA))
+    os.system('maxima -l ecl')
-- 
cgit v1.0-1-gd88e


From 88992bfae6d2aba0819b8cd67ae36621a5873b38 Mon Sep 17 00:00:00 2001
From: Michael Orlitzky <michael@orlitzky.com>
Date: Fri, 12 Nov 2021 20:51:28 -0500
Subject: Trac #32867: new spkg-configure.m4 for maxima.

This new spkg-configure.m4 checks for both a "maxima" executable, and
the usability of a "maxima" package by ECL. (The latter requires a
patched maxima until a new release with commit a0d7a43e523 is made.)
Notably absent for the moment is a version check, but feature testing
is moot until we decide what to do about matrixexp.patch.

This commit also moves the SAGE_MAXIMA_FAS variable handling from
ECL's spkg-configure.m4 to maxima's (where it would have belonged in
the first place, if maxima had an spkg-configure.m4 at the time.)
---
 build/pkgs/ecl/spkg-configure.m4    |  5 +----
 build/pkgs/maxima/spkg-configure.m4 | 35 +++++++++++++++++++++++++++++++++++
 pkgs/sage-conf/sage_conf.py.in      |  3 ++-
 3 files changed, 38 insertions(+), 5 deletions(-)
 create mode 100644 build/pkgs/maxima/spkg-configure.m4

diff --git a/build/pkgs/ecl/spkg-configure.m4 b/build/pkgs/ecl/spkg-configure.m4
index ae1e0ac..7dbcfa6 100644
--- a/build/pkgs/ecl/spkg-configure.m4
+++ b/build/pkgs/ecl/spkg-configure.m4
@@ -35,10 +35,7 @@ SAGE_SPKG_CONFIGURE([ecl], [
     AC_SUBST(SAGE_ECL_CONFIG, [$ECL_CONFIG])
   fi
 
-  # Maxima cannot yet be provided by the system, so we always use
+  # Kenzo cannot yet be provided by the system, so we always use
   # the SAGE_LOCAL path for now.
-  AC_SUBST(SAGE_MAXIMA_FAS, ['${prefix}'/lib/ecl/maxima.fas])
-
-  # Likewise for the optional Kenzo SPKG
   AC_SUBST(SAGE_KENZO_FAS, ['${prefix}'/lib/ecl/kenzo.fas])
 ])
diff --git a/build/pkgs/maxima/spkg-configure.m4 b/build/pkgs/maxima/spkg-configure.m4
new file mode 100644
index 00000000..df625fe
--- /dev/null
+++ b/build/pkgs/maxima/spkg-configure.m4
@@ -0,0 +1,35 @@
+SAGE_SPKG_CONFIGURE([maxima], [
+  SAGE_SPKG_DEPCHECK([ecl], [
+    dnl First check for the "maxima" executable in the user's PATH, because
+    dnl we still use pexpect to communicate with it in a few places. We pass
+    dnl the "-l ecl" flag here to ensure that the standalone executable also
+    dnl supports ECL.
+    AC_MSG_CHECKING(if "maxima -l ecl" works)
+    AS_IF([! maxima -l ecl -q -r 'quit();' \
+             >&AS_MESSAGE_LOG_FD 2>&AS_MESSAGE_LOG_FD], [
+      AC_MSG_RESULT(no)
+      sage_spkg_install_maxima=yes
+    ], [
+      AC_MSG_RESULT(yes)
+      dnl If we have the executable, check also for the ECL library.
+      AC_MSG_CHECKING([if ECL can "require" the maxima module])
+      AS_IF([ecl --eval "(require 'maxima)" --eval "(quit)" \
+               >&AS_MESSAGE_LOG_FD 2>&AS_MESSAGE_LOG_FD], [
+        AC_MSG_RESULT(yes)
+      ], [
+	AC_MSG_RESULT(no)
+	sage_spkg_install_maxima=yes
+      ])
+    ])
+  ])
+],[],[],[
+  # post-check
+  AS_IF([test x$sage_spkg_install_maxima = xyes], [
+    dnl Leaving this variable empty will tell sagelib to load
+    dnl the maxima library (within ECL) by name instead of by
+    dnl absolute path.
+    SAGE_MAXIMA_FAS='${prefix}'/lib/ecl/maxima.fas
+  ])
+  AC_SUBST(SAGE_MAXIMA_FAS, "${SAGE_MAXIMA_FAS}")
+])
+
diff --git a/pkgs/sage-conf/sage_conf.py.in b/pkgs/sage-conf/sage_conf.py.in
index b40cda6..8cda3eb 100644
--- a/pkgs/sage-conf/sage_conf.py.in
+++ b/pkgs/sage-conf/sage_conf.py.in
@@ -9,7 +9,8 @@ VERSION = "@PACKAGE_VERSION@"
 SAGE_LOCAL = "@prefix@"
 SAGE_ROOT = "@SAGE_ROOT@"
 
-# Delete this line if your ECL can load maxima without further prodding.
+# Set this to the empty string if your ECL can load maxima without
+# further prodding.
 MAXIMA_FAS = "@SAGE_MAXIMA_FAS@".replace('${prefix}', SAGE_LOCAL)
 
 # Delete this line if your ECL can load Kenzo without further prodding.
-- 
cgit v1.0-1-gd88e

